### PR TITLE
Allow input to be a Stream.

### PR DESCRIPTION
**Pull request by CodeMan99**
From conversation in issue #16.

The idea is that the first four bytes are read, then check file type. If not a zip, stop reading and return empty array. If it is a zip, buffer the stream and unzip as per normal.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -73,14 +75,29 @@ const extractFile = zip => new Promise((resolve, reject) => {
...
-		return Promise.reject(new TypeError(`Expected a Buffer, got ${typeof buf}`));
-	}
+module.exports = () => input => {
+	const processZip = buf => pify(yauzl.fromBuffer)(buf, {lazyEntries: true}).then(extractFile);
+
+	if (Buffer.isBuffer(input)) {
+		const type = fileType(input);
+
+		return type && type.ext === 'zip' ? processZip(input) : Promise.resolve([]);
+	} else if (isStream(input)) {
```
Comment (kevva) [inline index.js]: Don't need `else if` here since we're returning inside the previous `if` statement.

File: index.js [hunk 2]
```diff
@@ -73,14 +75,29 @@ const extractFile = zip => new Promise((resolve, reject) => {
...
+		const notAZipError = new Error('not-a-zip');
+		const firstChunkIsZip = new FirstChunkStream({chunkLength: 4}, (err, chunk, enc, cb) => {
+			const type = chunk && fileType(chunk);
+
+			cb(type && type.ext === 'zip' ? err : err || notAZipError, chunk);
+		});
+
+		return getStream.buffer(input.pipe(firstChunkIsZip)).then(processZip, error => {
+			if (error === notAZipError) {
```
Comment (kevva) [inline index.js]: `error` => `err`

File: index.js [hunk 3]
```diff
@@ -73,14 +75,29 @@ const extractFile = zip => new Promise((resolve, reject) => {
...
+			}
 
-	if (!fileType(buf) || fileType(buf).ext !== 'zip') {
-		return Promise.resolve([]);
+			throw error;
+		});
 	}
 
-	return pify(yauzl.fromBuffer)(buf, {lazyEntries: true}).then(extractFile);
+	return Promise.reject(new TypeError(`Expected a Buffer or Stream, got ${typeof input}`));
```
Comment (kevva) [inline index.js]: Don't move this to the bottom. I want to handle exceptions as early as possible.

**Review by kevva — CHANGES_REQUESTED** [review]

File: index.js (hunk 3 above)
Comment (CodeMan99) [reply index.js]: How is it different? Moving it to the top requires `if (!isStream(input) && !Buffer.isBuffer(input))` before it. Those two expressions (besides negation) are already directly in front of this line.

File: index.js (hunk 1 above)
Comment (CodeMan99) [reply index.js]: True, personally think that `else if` says "or" better, but I'll change it.

File: test.js [hunk 4]
```diff
@@ -43,6 +51,13 @@ test('return empty array if non-valid file is supplied', async t => {
 	t.is(files.length, 0);
 });
 
+test('return empty array if non-valid file stream is supplied', async t => {
+	const stream = fs.createReadStream(__filename);
+	const files = await m()(stream);
+
+	t.is(files.length, 0);
```
Comment (CodeMan99) [inline test.js:59]: Should probably check the readable state of the stream here, do you know what is appropriate? I am thinking `stream._readableState.flowing === false` and `stream._readableState.ended === false`.

**Comment by CodeMan99** [comment]
@kevva did you go on vacation? Would love to have some guidance on this if you are available.

**Comment by CodeMan99** [comment]
@kevva Would love to get this PR ready for merge while the information is still fresh in my mind.

Also, it might be worth separating the "streaming file-type" part since something similar will likely need done for the `decompress` module.

File: index.js (hunk 3 above)
Comment (CodeMan99) [reply index.js]: @kevva Still eagerly awaiting your response.

Comment (CodeMan99) [reply index.js]: Implemented this suggestion despite not really understanding the reasoning.

File: test.js (hunk 4 above)
Comment (CodeMan99) [reply test.js:59]: implemented as described.

**Comment by CodeMan99** [comment]
Closing because questions have gone unanswered.

**Comment by kevva** [comment]
@CodeMan99, I don't see the point in allowing a stream as input if we're going to buffer it anyway. That misses the whole point of streaming. Unless `yauzl` start supporting streams, or if there is a alternative module that supports it, I'm not going merge this.

**Pull request closed without merging**